tests/Congress.test.ts

```diff
@@ -1127,573 +1235,721 @@ describe("Test for the creation a proposal and the voting", () => {
+//         assert.strictEqual(validators.length, validators_simulation.length);
+//         assert.deepStrictEqual(
+//             validators.map((m) => m.address.toString()),
+//             validators_simulation.map((m) => m.toString())
+//         );
+//     });
+
+//     it("Create a block 4 - Dummy", async () => {
+//         const new_block = await createDummyBlock(4);
+//         const validators_simulation = block_manager.getValidators(block_manager.getLastBlockHeight());
+//         assert.strictEqual(new_block.header.height.toString(), "4");
+//         assert.strictEqual(new_block.header.validators.length, 8);
+//         assert.strictEqual(validators_simulation.length, new_block.header.validators.length);
+
+//         const validators = await boa_client.getAllValidators(block_manager.getLastBlockHeight());
+//         assert.strictEqual(validators.length, validators_simulation.length);
+//         assert.deepStrictEqual(
+//             validators.map((m) => m.address.toString()),
+//             validators_simulation.map((m) => m.toString())
+//         );
+//     });
+
+//     it("Create a block 5 - Slash one validator", async () => {
+//         block_manager.removeValidator(ValidatorKey.keys(0));
+//         const new_block = await createDummyBlock(5);
+//         const validators_simulation = block_manager.getValidators(block_manager.getLastBlockHeight());
+//         assert.strictEqual(new_block.header.height.toString(), "5");
+//         assert.strictEqual(new_block.header.validators.length, 8);
+//         assert.strictEqual(validators_simulation.length, new_block.header.validators.length);
+
+//         const validators = await boa_client.getAllValidators(block_manager.getLastBlockHeight());
+//         assert.strictEqual(validators.length, validators_simulation.length);
+//         assert.deepStrictEqual(
+//             validators.map((m) => m.address.toString()),
+//             validators_simulation.map((m) => m.toString())
+//         );
+//     });
+
+//     it("Create a block 6 - Dummy", async () => {
+//         const new_block = await createDummyBlock(6);
+//         const validators_simulation = block_manager.getValidators(block_manager.getLastBlockHeight());
+//         assert.strictEqual(new_block.header.height.toString(), "6");
+//         assert.strictEqual(new_block.header.validators.length, 7);
+//         assert.strictEqual(validators_simulation.length, new_block.header.validators.length);
+
+//         const validators = await boa_client.getAllValidators(block_manager.getLastBlockHeight());
+//         assert.strictEqual(validators.length, validators_simulation.length);
+//         assert.deepStrictEqual(
+//             validators.map((m) => m.address.toString()),
+//             validators_simulation.map((m) => m.toString())
+//         );
+//     });
+// });
+
+// describe("Test of Proposal API", function () {
```

Comment:
This is a test code made using BlockManager.
I made it similar after looking at this test scenario.
I want to change it to this.
And please find out the cause of the error.



``` TypeScript
describe("Test of Proposal API", function () {
    this.timeout(5000);
    const agora_addr: URL = new URL("http://localhost:2862");
    const stoa_addr: URL = new URL("http://localhost:3862");
    const stoa_private_addr: URL = new URL("http://localhost:4862");
    const votera_addr: URL = new URL("http://127.0.0.1:1337/");
    const client = new TestClient();

    let stoa_server: TestStoa;
    let agora_server: TestAgora;
    let testDBConfig: IDatabaseConfig;
    let gecko_server: TestGeckoServer;
    let gecko_market: CoinGeckoMarket;
    let coinMarketService: CoinMarketService;
    let votera_server: TestVoteraServer;
    let votera_service: VoteraService;

    let block_manager: BlockManager;
    let gen_keypair: KeyPair;
    let boa_client: BOAClient;
    let utxo_provider: UTXOProvider;
    let proposer_utxo_provider: UTXOProvider;
    let tx_hash_proposal_fee: Hash;
    let tx_hash_proposal: Hash;

    let proposal_key_pair: KeyPair;
    let proposal_fee_destination: PublicKey;

    let ballet_yes: Hash;
    let ballet_no: Hash;

    async function createDummyBlock(expected_block_height: number) {
        const new_block = block_manager.saveBlock([], []);
        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: new_block });
        await block_manager.waitFor(block_manager.getLastBlockHeight(), boa_client);
        assert.strictEqual(JSBI.toNumber(await boa_client.getBlockHeight()), block_manager.getLastBlockHeight());
        assert.strictEqual(block_manager.getLastBlockHeight(), expected_block_height);
    }

    before("Bypassing middleware check", () => {
        FakeBlacklistMiddleware.assign();
    });

    before("Wait for the package libsodium to finish loading", async () => {
        if (!SodiumHelper.isAssigned()) SodiumHelper.assign(new BOASodium());
        await SodiumHelper.init();
    });

    before("Start a fake Agora", () => {
        return new Promise<void>((resolve, reject) => {
            agora_server = new TestAgora(agora_addr.port, [], resolve);
        });
    });

    before("Start a fake TestCoinGeckoServer", () => {
        return new Promise<void>(async (resolve, reject) => {
            gecko_server = new TestGeckoServer("7876", market_cap_sample_data, market_cap_history_sample_data, resolve);
            gecko_market = new CoinGeckoMarket(gecko_server);
        });
    });

    before("Start a fake TestCoinGecko", () => {
        coinMarketService = new CoinMarketService(gecko_market);
    });

    before("Start a fake votera Server and Service", () => {
        return new Promise<void>(async (resolve, reject) => {
            votera_server = new TestVoteraServer(1337, votera_addr, resolve);
            votera_service = new VoteraService(votera_addr);
        });
    });

    before("Create TestStoa", async () => {
        testDBConfig = await MockDBConfig();
        stoa_server = new TestStoa(testDBConfig, agora_addr, stoa_addr.port, votera_service, coinMarketService);
        await stoa_server.createStorage();
    });

    before("Start TestStoa", async () => {
        await stoa_server.start();
        await stoa_server.voteraService?.stop();
    });

    after("Stop Stoa and Agora server instances", async () => {
        await stoa_server.ledger_storage.dropTestDB(testDBConfig.database);
        await stoa_server.stop();
        await votera_server.stop();
        await gecko_server.stop();
        await agora_server.stop();
    });

    before("Create component", async () => {
        block_manager = new BlockManager();
        gen_keypair = KeyPair.fromSeed(new SecretKey("SDN7BBGE6Z6OQM3K4PACLTZUJ5QX4AY4QPDQ2JJ2JCFWCG2OIYYALIRY"));
        boa_client = new BOAClient(stoa_addr.toString(), agora_addr.toString());
        utxo_provider = new UTXOProvider(gen_keypair.address, boa_client);

        proposal_key_pair = ValidatorKey.keys(0);
        proposal_fee_destination = new PublicKey("boa1xrgq6607dulyra5r9dw0ha6883va0jghdzk67er49h3ysm7k222ruhh7400");
        proposer_utxo_provider = new UTXOProvider(proposal_key_pair.address, boa_client);
    });

    it("0. Test of the path /block_externalized", async () => {
        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: Block.reviver("", sample_data[0]) });
        await delay(500);
    });

    it("1. Distribute genesis coin", async () => {
        const utxos = await utxo_provider.getUTXO(BOA(1000_000 * 100 + 10));
        const builder = new TxBuilder(gen_keypair);
        utxos.forEach((m) => {
            builder.addInput(m.utxo, m.amount);
        });
        iota(100).forEach((idx: number) => {
            builder.addOutput(ValidatorKey.keys(idx).address, BOA(1000_000));
        });

        const tx = builder.sign(OutputType.Payment, BOA(10));
        const new_block = block_manager.saveBlock([tx], []);

        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: new_block });
        await block_manager.waitFor(block_manager.getLastBlockHeight(), boa_client);
        assert.strictEqual(JSBI.toNumber(await boa_client.getBlockHeight()), block_manager.getLastBlockHeight());
        assert.strictEqual(block_manager.getLastBlockHeight(), 1);
    });

    it("2. Create a proposal fee data", async () => {
        const proposal_fee_data = new ProposalFeeData("Votera", "469008972006");
        const proposal_fee = BOA(1_000);
        const buffer = new SmartBuffer();
        proposal_fee_data.serialize(buffer);
        const payload = buffer.toBuffer();

        const utxos = await proposer_utxo_provider.getUTXO(proposal_fee);
        const builder = new TxBuilder(proposal_key_pair);
        utxos.forEach((m) => {
            builder.addInput(m.utxo, m.amount);
        });
        builder.addOutput(proposal_fee_destination, proposal_fee);
        builder.assignPayload(payload);

        const tx = builder.sign(OutputType.Payment, BOA(1));

        tx_hash_proposal_fee = hashFull(tx);
        console.log(tx_hash_proposal_fee.toString());

        assert.deepStrictEqual(
            tx_hash_proposal_fee.toString(),
            "0x33f661e8d38ae9a2f096f50822306e5ab06a7d000a9d52f7ff476b2c533d145ebab740d0c296a438a5782405e15365286f49e1e2d6db158334ed0a7b24af7470"
        );

        assert.deepStrictEqual(
            proposal_key_pair.address.toString(),
            "boa1xpvald2ydpxzl9aat978kv78y5g24jxy46mcnl7munf4jyhd0zjrc5x62kn"
        );

        const new_block = block_manager.saveBlock([tx], []);

        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: new_block });
        await block_manager.waitFor(block_manager.getLastBlockHeight(), boa_client);
        assert.strictEqual(JSBI.toNumber(await boa_client.getBlockHeight()), block_manager.getLastBlockHeight());
        assert.strictEqual(block_manager.getLastBlockHeight(), 2);
    });

    it("3. Create a proposal data", async () => {
        const validators = block_manager.getValidators();
        const vote_cost = Amount.add(
            Amount.make(Utils.FEE_RATE * Transaction.getEstimatedNumberOfBytes(1, 2, 285)),
            TxPayloadFee.getFeeAmount(285)
        );
        const total_vote_cost = Amount.multiply(vote_cost, validators.length);

        const proposal_data = new ProposalData(
            "Votera",
            ProposalType.Fund,
            "469008972006",
            "Make better world!",
            JSBI.BigInt(3),
            JSBI.BigInt(6),
            new Hash(Buffer.alloc(Hash.Width)),
            BOA(100_000).value,
            BOA(1_000).value,
            total_vote_cost.value,
            tx_hash_proposal_fee,
            proposal_key_pair.address,
            proposal_fee_destination
        );
        const buffer = new SmartBuffer();
        proposal_data.serialize(buffer);
        const payload = buffer.toBuffer();

        const tx_total_fee = Amount.add(
            Amount.make(
                Utils.FEE_RATE * Transaction.getEstimatedNumberOfBytes(1, validators.length + 1, payload.length)
            ),
            TxPayloadFee.getFeeAmount(payload.length)
        );
        const send_amount = Amount.add(total_vote_cost, tx_total_fee);
        const utxos = await proposer_utxo_provider.getUTXO(send_amount);

        const builder = new TxBuilder(proposal_key_pair);
        utxos.forEach((m) => {
            builder.addInput(m.utxo, m.amount);
        });
        validators.forEach((v) => {
            builder.addOutput(v, vote_cost);
        });
        builder.assignPayload(payload);

        const tx = builder.sign(OutputType.Payment, tx_total_fee);

        tx_hash_proposal = hashFull(tx);
        console.log(tx_hash_proposal.toString());

        const new_block = block_manager.saveBlock([tx], []);

        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: new_block });
        await block_manager.waitFor(block_manager.getLastBlockHeight(), boa_client);
        assert.strictEqual(JSBI.toNumber(await boa_client.getBlockHeight()), block_manager.getLastBlockHeight());
        assert.strictEqual(block_manager.getLastBlockHeight(), 3);
    });

    it("Start votera service for syncing proposal's meta information", async () => {
        await stoa_server.voteraService?.start(stoa_server, 2);
        await delay(1000);

        await stoa_server.voteraService?.stop();
    });

    it("4. Vote [ YES ]", async () => {
        // The KeyPair of the validator
        const validator_key = ValidatorKey.keys(1);
        const app_name = "Votera";
        const proposal_id = "469008972006";
        let vote = new Vote(boa_client, block_manager, validator_key, app_name, proposal_id, BallotData.YES, 100, 13);
        const tx = await vote.CreateVote();
        assert.ok(tx !== undefined);

        const new_block = block_manager.saveBlock([tx], []);

        ballet_yes = hashFull(tx);

        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: new_block });
        await block_manager.waitFor(block_manager.getLastBlockHeight(), boa_client);
        assert.strictEqual(JSBI.toNumber(await boa_client.getBlockHeight()), block_manager.getLastBlockHeight());
        assert.strictEqual(block_manager.getLastBlockHeight(), 4);
    });

    it("5. Vote [ No ]", async () => {
        // The KeyPair of the validator
        const validator_key = ValidatorKey.keys(1);
        const app_name = "Votera";
        const proposal_id = "469008972006";
        let vote = new Vote(boa_client, block_manager, validator_key, app_name, proposal_id, BallotData.NO, 100, 13);
        const tx = await vote.CreateVote();
        assert.ok(tx !== undefined);

        ballet_no = hashFull(tx);

        const new_block = block_manager.saveBlock([tx], []);

        const block_url = URI(stoa_private_addr).directory("block_externalized").toString();
        await client.post(block_url, { block: new_block });
        await block_manager.waitFor(block_manager.getLastBlockHeight(), boa_client);
        assert.strictEqual(JSBI.toNumber(await boa_client.getBlockHeight()), block_manager.getLastBlockHeight());
        assert.strictEqual(block_manager.getLastBlockHeight(), 5);
    });

    it("6-14 Create a dummy block", async () => {
        await createDummyBlock(6);
        await createDummyBlock(7);
        await createDummyBlock(8);
        await createDummyBlock(9);
        await createDummyBlock(10);
        await createDummyBlock(11);
        await createDummyBlock(12);
        await createDummyBlock(13);
        await createDummyBlock(14);
    });

    it("Test for path /proposal/voting-details/:proposal_id", async () => {
        const uri = URI(stoa_addr).directory("/proposal/voting-details").filename("469008972006");
        const response = await client.get(uri.toString());
        console.log(JSON.stringify(response.data));
        assert.strictEqual(response.data.length, 2);
        assert.strictEqual(response.data[0].address, ValidatorKey.keys(1).address.toString());
        assert.strictEqual(response.data[1].address, ValidatorKey.keys(1).address.toString());
        const ballot_data_yes = response.data.find((m: any) => m.hash === ballet_yes.toString());
        assert.ok(ballot_data_yes !== undefined);
        assert.strictEqual(ballot_data_yes.ballot_answer, "Yes");

        const ballot_data_no = response.data.find((m: any) => m.hash === ballet_no.toString());
        assert.ok(ballot_data_no !== undefined);
        assert.strictEqual(ballot_data_no.ballot_answer, "No");
    });

    it("Test for path /validator/ballot/:address", async () => {
        const uri = URI(stoa_addr)
            .directory("/validator/ballot")
            .filename("boa1xrvald3zmehvpcmxqm0kn6wkaqyry7yj3cd8h975ypzlyz00sczpzhsk308");
        const response = await client.get(uri.toString());
        console.log(JSON.stringify(response.data));
        assert.deepStrictEqual(response.data.length, 2);
        assert.deepStrictEqual(response.data[0].proposal_id, "469008972006");
        assert.deepStrictEqual(response.data[1].proposal_id, "469008972006");
        assert.deepStrictEqual(response.data[0].proposal_type, "Fund");
        assert.deepStrictEqual(response.data[1].proposal_type, "Fund");

        const ballot_data_yes = response.data.find((m: any) => m.tx_hash === ballet_yes.toString());
        assert.ok(ballot_data_yes !== undefined);
        assert.strictEqual(ballot_data_yes.ballot_answer, "Yes");

        const ballot_data_no = response.data.find((m: any) => m.tx_hash === ballet_no.toString());
        assert.ok(ballot_data_no !== undefined);
        assert.strictEqual(ballot_data_no.ballot_answer, "No");
    });
});
```

Reply:
TestVoteraServer has been modified like this.
```
export class TestVoteraServer {
    private server: http.Server;

    private app: express.Application;

    constructor(port: number, votera_endpoint: URL, done: () => void) {
        this.app = express();
        this.server = http.createServer(this.app);
        this.server.listen(port, () => {
            done();
        });
        this.app.get("/votera-proposal/469008972006", (req: Request, res: Response) => {
            const response = {
                proposalId: "469008972006",
                proposer_address: "boa1xpvald2ydpxzl9aat978kv78y5g24jxy46mcnl7munf4jyhd0zjrc5x62kn",
                name: "Make better world!",
                type: "BUSINESS",
                status: "closed",
                votePeriod: { begin: "2021-07-26", end: "2021-08-02" },
                createdAt: "2021-07-23T04:49:26.634Z",
                description: "Description Make better world!",
                proposal_fee_tx_hash:
                    "0x33f661e8d38ae9a2f096f50822306e5ab06a7d000a9d52f7ff476b2c533d145ebab740d0c296a438a5782405e15365286f49e1e2d6db158334ed0a7b24af7470",
                vote_fee: 19.17672,
                fundingAmount: 100000,
                voting_start_height: 3,
                voting_end_height: 6,
                tx_hash_vote_fee:
                    "0xba747aa032f83fa6754efee9b259ac181eaa1f657395cb09b9f79a2d198fd0ce83e441c8a1aba51bd29b400b8bde27bacc22bef8e1f5261e8e545102afb96f4d",
                assessResult: {
                    average: 7,
                    nodeCount: 2,
                    completeness: 6,
                    realization: 6.5,
                    profitability: 7,
                    attractiveness: 7.5,
                    expansion: 8,
                },
                assessPeriod: { begin: "2021-08-18", end: "2021-08-18" },
                creator: { username: "test" },
                attachment: [
                    {
                        id: "61f6724251k789",
                        name: "Make the world better",
                        url: "https://s3.ap-northeast-2.amazonaws.com/com.kosac.defora.beta.upload-image/BOASCAN_Requirements_Documentation_Version1_0_EN_copy_fb69a8a7d5.pdf",
                        mime: "application/pdf",
                        doc_hash: "5b5073302c8570a269a5d028cc256d80b7d5d22aaa05e279fac7ced94d7df7c9",
                    },
                ],
            };
            return res.status(200).send(response);
        });
        this.app.get("/votera-proposal/469008972001", (req: Request, res: Response) => {
            const response = {
                proposalId: "469008972001",
                proposer_address: "boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh",
                name: "Make better world!",
                type: "BUSINESS",
                status: "closed",
                votePeriod: { begin: "2021-07-26", end: "2021-08-02" },
                createdAt: "2021-07-23T04:49:26.634Z",
                description: "Description Make better world!",
                proposal_fee_tx_hash:
                    "0x11c6b0395c8e1716978c41958eab84e869755c09f7131b3bbdc882a647cb3f2c46c450607c6da71d34d1eab28fbfdf14376b444ef46ed1d0a7d2237ab430ebf5",
                vote_fee: 100,
                fundingAmount: 100000,
                voting_start_height: 3,
                voting_end_height: 6,
                tx_hash_vote_fee:
                    "0x8b6a2e1ecc3616ad63c73d606c4019407ebfd06a122519e7bd88d99af92d19d9621323d7c2e68593053a570522b6bc8575d1ee45a74ee38726f297a5ce08e33d",
                assessResult: {
                    average: 7,
                    nodeCount: 2,
                    completeness: 6,
                    realization: 6.5,
                    profitability: 7,
                    attractiveness: 7.5,
                    expansion: 8,
                },
                assessPeriod: { begin: "2021-08-18", end: "2021-08-18" },
                creator: { username: "test" },
                attachment: [
                    {
                        id: "61f6724251k789",
                        name: "Make the world better",
                        url: "https://s3.ap-northeast-2.amazonaws.com/com.kosac.defora.beta.upload-image/BOASCAN_Requirements_Documentation_Version1_0_EN_copy_fb69a8a7d5.pdf",
                        mime: "application/pdf",
                        doc_hash: "5b5073302c8570a269a5d028cc256d80b7d5d22aaa05e279fac7ced94d7df7c9",
                    },
                ],
            };
            return res.status(200).send(response);
        });
    }

    // Shut down
    public stop(): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            this.server.close((err?) => {
                err === undefined ? resolve() : reject(err);
            });
        });
    }
}
```

Reply:
@ahmedrns The validator can vote several times.
The final ballot should be reflected in the tally.
However, on the test code, the second ballot is blank.
It is an issue that must be solved.

Reply:
@MichaelKim20 I have added some additional test cases in Proposal API test, ballots with max sequence will be displayed